src/python/pants/option/option_types_test.py

```diff
@@ -104,45 +107,73 @@ class MySubsystem(MyBaseSubsystem):
     assert my_subsystem.dyn_prop == transform_opt(option_value)
 
 
-def test_other_options():
-    class MySubsystem(Subsystem):
+def test_other_options() -> None:
+    class MyBaseSubsystem(Subsystem):
         def __init__(self):
             self.options = SimpleNamespace()
-            self.options.dict_opt = {"key1": "val1"}
             self.options.enum_opt = MyEnum.Val2
             self.options.optional_enum_opt = MyEnum.Val2
+            self.options.dyn_enum_opt = MyEnum.Val2
             self.options.enum_list_opt = [MyEnum.Val2]
+            self.options.dyn_enum_list_opt = [MyEnum.Val2]
             self.options.defaultless_enum_list_opt = [MyEnum.Val2]
+            self.options.dict_opt = {"key1": "val1"}
             self.options.args = ["--arg1"]
 
-        dict_prop = DictOption[Any]("--dict-opt", help="")
         enum_prop = EnumOption("--enum-opt", default=MyEnum.Val1, help="")
+        dyn_enum_prop = EnumOption(
+            "--dyn-enum-opt",
+            default=lambda cls: cls.dyn_default,
+            enum_type=MyEnum,
+            help=lambda cls: f"{cls.dyn_help}",
```

Comment:
This is kind of a bummer, in order to shut `mypy` up I had to use a `f`-string. I would imagine the help is static with a few dynamic values, so it would match this case.

Let's keep an eye-out and we can revisit if it gets annoying.